python/tank/descriptor/io_descriptor/appstore.py

```diff
@@ -538,6 +538,165 @@ def get_latest_cached_version(self, constraint_pattern=None):
         log.debug("Latest cached version resolved to %r" % desc)
         return desc
 
+    def _filter_versions_by_label(self, all_versions):
+        """
+        Filter cached versions by label if one is specified.
+
+        :param all_versions: Dictionary of {version_string: path}
+        :returns: List of version strings that match the label criteria
+        """
+        if not self._label:
+            return list(all_versions.keys())
+
+        version_numbers = []
+        log.debug(f"Filtering cached versions by label '{self._label}'...")
+        for version_str, path in all_versions.items():
+            metadata = self.__load_cached_app_store_metadata(path)
+            try:
+                tags = [x["name"] for x in metadata["sg_version_data"]["tags"]]
+                if self.__match_label(tags):
+                    version_numbers.append(version_str)
+            except Exception as e:
+                log.debug(
+                    f"Could not determine label metadata for {path}. Ignoring. Details: {e}"
+                )
+        return version_numbers
+
+    def _filter_versions_by_app_store(self, version_numbers, sg_matching_records):
+        """
+        Filter versions to only those available in the App Store.
+
+        :param version_numbers: List of version strings
+        :param sg_matching_records: List of Shotgun version records from App Store
+        :returns: List of version strings that exist in App Store, or original list if no records provided
+        """
+        if not sg_matching_records:
+            return version_numbers
+
+        app_store_versions = {rec["code"] for rec in sg_matching_records}
+        return [v for v in version_numbers if v in app_store_versions]
+
+    def _sort_versions_descending(self, version_numbers):
+        """
+        Sort version numbers in descending order (newest first).
+
+        :param version_numbers: List of version strings
+        :returns: Sorted list with newest versions first
+        """
+        try:
+            sorted_versions = []
+            for v in version_numbers:
+                inserted = False
+                for i, existing in enumerate(sorted_versions):
+                    if is_version_newer(v, existing):
+                        sorted_versions.insert(i, v)
+                        inserted = True
+                        break
+                if not inserted:
+                    sorted_versions.append(v)
+            return sorted_versions
```

Comment:
Can we use `sorted` with a `key` parameter instead here?
I think we do that already in core dcc version comparison